fastcrypto-zkp/examples/blake2s_demo.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use ark_bls12_381::{Bls12_381, Fr};
+use ark_ff::ToConstraintField;
+use ark_groth16::{create_random_proof, generate_random_parameters};
+use ark_std::rand::Rng;
+
+use ark_crypto_primitives::{
+    prf::blake2s::constraints::Blake2sGadget, prf::blake2s::Blake2s as B2SPRF, PRFGadget, PRF,
+};
+use ark_relations::r1cs::{
+    ConstraintSynthesizer, ConstraintSystem, ConstraintSystemRef, SynthesisError,
+};
+use blake2::{digest::Digest, Blake2s256};
+
+use ark_r1cs_std::prelude::*;
+use fastcrypto_zkp::verifier::{process_vk_special, verify_with_processed_vk};
+
+#[derive(Clone, Copy, Debug)]
+struct Blake2sCircuit {
+    input: [u8; 32],
+    blake2_seed: [u8; 32],
+    pub expected_output: [u8; 32],
+}
+
+impl Blake2sCircuit {
```

Comment:
Awesome, 
Given `some_public_string`, `h = Blake2s(seed, some_public_string)` and a `ZKP(I know secret of h)`, this is literally a signature HMAC scheme proven in ZK, not just a hash, right?

Reply:
It depends what you mean by "this", in that the exact code you're looking at in this PR sets the variable `some_public_string` as a (hidden) witness rather than a public input. But it would be an easy modification to do what you intend to, yes.

Reply:
@huitseeker what's the point if both the `seed` and `some_public_string` are hidden? I'd expect two reasonable examples. One for which you prove you know the preimage (no public inputs) (not sure if the seed is required here).

And another that requires the `seed` to be the secret (witness) and `some_public_string` be a public input. 

Both examples have different applications, and hence we probably need them both.

Reply:
Please do only if it's easy to modify. I used at least both of them in many pitching sessions in the past next to each other to explain ie how easy it is to transform (and why a small transformation is a completely different cool application)

Reply:
No, your feedback makes total sense, I've now modified the example to present:
> seed, some_public_string (seed is witness, the other is public input)
which is closer to the objective of the example. I was just explaining why one would not want to present *all three examples* jointly in this file.